---
title: "ihs"
author: "Murray Cadzow"
date: "18/03/2015"
output: html_document
---

```{r, fig.height=9, fig.width=12}
library(dplyr)
library(ggplot2)
library(scales)
library(rehh)
library(qqman)
library(knitr)
library(readr)
dir="/run/user//1001//gvfs/smb-share:server=biocldap,share=scratch/merrimanlab/murray/working_dir/extract/ihh_ihs/"
POP=""
ihs=data.frame()
s=data.frame()
make_graph=function(ihs,s){
  ggplot(data = ihs, aes(x=POSITION, y=iHS), ylim=c(min(ihs$iHS -0.5), max(ihs$iHS + 0.5)))   + facet_wrap(~CHR, scales = "free_x")  + geom_hline(aes(yintercept= q1, colour ="quantile"), data=s,)  + geom_hline(aes(yintercept= q2, colour ="quantile"), data=s)  +geom_hline(aes(yintercept= m, colour="mean"), data=s) + scale_colour_manual("",breaks=c("mean","quantile"),values=c("blue","red")) + scale_x_continuous( xlab("Chromosome Position (Mbp)")) + ylab("iHS") + ggtitle(paste0(POP," iHS by Chromosome")) + theme( plot.background= element_rect(colour="black",fill=NA), legend.position= c(0.75, 0.12)) + theme_bw() + geom_hex()#+ geom_point(shape = 16, colour= alpha("black", 1/5))
  }
#setwd("/run/user//1001/gvfs/smb-share:server=biocldap,share=scratch/merrimanlab/murray/working_dir/extract/ihh_ihs/")


#for( POP in c("AXIOM","OMNI","CEU","CHB","CHS","GBR","YRI")){
#  print(POP)
read_ihs=function(POP){  
  if(!file.exists(paste0(dir,POP,"_ihs.RData"))){
    ihh=data.frame()
    for( i in 1:22){
      if(file.exists(file = paste0(dir,POP,"chr",i,".ihh"))){
        ihh=rbind(ihh,read.table(file = paste0(dir,POP,"chr",i,".ihh"), header=TRUE))
      } else {
        print(paste0(POP,"chr",i,".ihh is missing"))
      }
    }
    
    ihs = ihh2ihs(ihh)
    ihs = ihs$res.ihs
    ihs = ihs[!is.na(ihs$iHS),]
    write.table(ihs, file=paste0(dir,POP,"_ihs.txt"), row.names=FALSE, quote=FALSE,col.names = TRUE, sep="\t")
  }else{
    ihs = read_delim(file=paste0(dir,POP,"_ihs.txt"), col_names=TRUE, delim="\t")
  }
  #print(names(ihs))
  #print(head(ihs))
  
  #plot(make_graph(TD,s))
  return(ihs)
  }

plot_manhattan=function(ihs){
  manhattan(x = ihs[!is.na(ihs$iHS),], chr = "CHR", bp = "POSITION", p = "Pvalue", main=POP)
}


ihs_stats = function(ann){
  print(
    as.data.frame(
      ann1 %>% group_by(CHR) %>% summarise(m=mean(iHS), sd(iHS), min(iHS), max(iHS), q1 = quantile(iHS, 0.01), q2 = quantile(iHS, 0.99))
    )
  )
  print("Top SNPs")
  print(
    as.data.frame(
  ann1 %>% group_by(CHR) %>% filter(max(abs(iHS))== abs(iHS)) %>% arrange(CHR)
    )
  )
  print("Top Genes")
  print(
    as.data.frame(
  ann1 %>% group_by(CHR) %>% filter(gene != "NA") %>% filter(max(abs(iHS))== abs(iHS)) %>% arrange(CHR)
    )
  )
}

df=data.frame()
for(i in 1:22){
  df=rbind(df,read_delim(file=paste0("/run/user//1001/gvfs/smb-share:server=biocldap,share=scratch/merrimanlab/murray/working_dir/extract/fixed_vcf/snpEff_classic/rs_genes",i,".txt"), delim="\t", col_names=c("chr","start","end","rs_ids","gene")) )
}

annotate=function(ihs){
  #ihs pos == df end pos
  #ann = merge(ihs, df, by.x=c("CHR","POSITION"), by.y=c("chr", "end"), all.x = TRUE, sort = FALSE)
  ann = left_join(ihs, df, by=c("CHR" ="chr", "POSITION"= "end"), sort = FALSE)

  ann1 = ann %>% arrange(desc(abs(iHS)))
  ann1$rank = as.numeric(rownames(ann1))
  ann1$rel_rank = as.numeric(rownames(ann1))/NROW(ann1)
  
  genes=c("ABCG2", "SLC2A9", "PPARGC1A","PDZK1","GCKR","RREB1","SLC17A3","SLC16A9", "SLC22A11","SLC22A12", "INHBE")
  print("rank statistics by gene")
  plot(ann1 %>% filter(gene %in% genes) %>% group_by(gene) %>% mutate(gene_num = paste(gene,length(gene))) %>%
         ggplot(. , aes(x=gene, y=rank), ylim=c(0,6e6)) + 
         geom_boxplot() + 
         geom_jitter() +
         facet_wrap(~gene_num, scales = "free_x")+
         theme_bw() + 
         ggtitle("Old Kottgen Genes") +
         theme(axis.text.x=element_blank(),axis.ticks=element_blank())
       )
  print("iHS statistics by gene")
  plot(ann1 %>% filter(gene %in% genes) %>% group_by(gene) %>% mutate(gene_num = paste(gene,length(gene))) %>%
         ggplot(. , aes(x=gene, y=iHS), ylim=c(0,6e6)) + 
         geom_boxplot() +
         geom_jitter() +
         facet_wrap(~gene_num, scales = "free_x")+
         theme_bw() + 
         ggtitle("Old Kottgen Genes iHS") +
         theme(axis.text.x=element_blank(),axis.ticks=element_blank())
       )
  
  genes_new=c("TRIM46", "INHBB", "ORC4L", "SFMBT1","TMEM171", 
              "VEGFA","MLXIPL","PRKAG2","STC1","HNF4G","ASAH2", 
              "LTBP3", "PTPN11", "NRG4", "IGF1R", "NFAT5", "MAF",
              "HLF", "BCAS3", "QRICH2")
  print("rank statistics by gene")
  plot(ann1 %>% filter(gene %in% genes_new) %>% group_by(gene) %>% mutate(gene_num = paste(gene,length(gene))) %>%
         ggplot(. , aes(x=gene, y=rank), ylim=c(0,6e6)) + 
         geom_boxplot() + 
         geom_jitter() +
         facet_wrap(~gene_num, scales = "free_x")+
         theme_bw() +
         ggtitle("New Kottgen Genes") +
         theme(axis.text.x=element_blank(),axis.ticks=element_blank())
       )
  print("iHS statistics by gene")
  plot(ann1 %>% filter(gene %in% genes_new) %>% group_by(gene) %>% mutate(gene_num = paste(gene,length(gene))) %>%
         ggplot(. , aes(x=gene, y=iHS), ylim=c(0,6e6)) + 
         geom_boxplot() + 
         geom_jitter() +
         facet_wrap(~gene_num, scales = "free_x")+
         theme_bw() +
         ggtitle("New Kottgen Genes iHS") +
         theme(axis.text.x=element_blank(),axis.ticks=element_blank())
       )
  
  print(as.data.frame(
    ann1 %>% 
      filter(gene %in% c(genes, genes_new)) %>% 
      group_by(gene) %>% 
      summarise(min=min(rank), 
                firstQ=quantile(rank, 0.25),
                median=median(rank), 
                mean=mean(rank),
                thirdQ=quantile(rank, 0.75), 
                max=max(rank),
                numSnps=length(rank)
      )
  ))
  a<- ann1 %>%  
  group_by(gene) %>% 
  summarise(min_rank=min(rank),
            firstQu=quantile(rank, 0.25),
            med=median(rank), 
            me=mean(rank), 
            thirdQu=quantile(rank, 0.75), 
            max_rank=max(rank), 
            len=length(rank), 
            v=var(rank)
            ) %>% 
  arrange(me, v)
  print(head(a, n=20))
  return(ann1)
}
```


---

AXIOM

---

```{r axiom, echo=FALSE, cache=FALSE}
axiom<-read_ihs(POP="AXIOM")
ann1 = annotate(axiom)
ihs_stats(ann1)
rm(axiom, ann1)
```

---

OMNI

---

```{r omni, echo=FALSE, cache=FALSE}
omni<-read_ihs(POP="OMNI")
ann1 = annotate(omni)
ihs_stats(ann1)
rm(omni,ann1)
```


---

CEU

---

```{r ceu, echo=FALSE, cache=FALSE}
ceu<-read_ihs(POP="CEU")
ann1 = annotate(ceu)
ihs_stats(ann1)
rm(ceu)
```

---

CHB

---

```{r chb, echo=FALSE, cache=FALSE}
chb<-read_ihs(POP="CHB")
ann1 = annotate(chb)
ihs_stats(ann1)
rm(chb)
```

---

CHS

---

```{r chs, echo=FALSE, cache=FALSE}
chs<-read_ihs(POP="CHS")
ann1 = annotate(chs)
ihs_stats(ann1)
rm(chs)
```

---

GBR

---

```{r gbr, echo=FALSE, cache=FALSE}
gbr<-read_ihs(POP="GBR")
ann1 = annotate(gbr)
ihs_stats(ann1)
rm(gbr)
```

---

YRI

---

```{r yri, echo=FALSE, cache=FALSE}
yri<-read_ihs(POP="YRI")
ann1 = annotate(yri)
ihs_stats(ann1)
rm(yri)
```